pkg/quarterdeck/middleware/ratelimiter.go

```diff
@@ -0,0 +1,117 @@
+package middleware
+
+import (
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/rotationalio/ensign/pkg/quarterdeck/api/v1"
+	"github.com/rotationalio/ensign/pkg/quarterdeck/config"
+	"golang.org/x/time/rate"
+)
+
+/*
+IPRateLimiter is an IP address based limiter that controls how frequently requests
+can be made from a single IP address.
+limit: represents the number of tokens that can be added to the token bucket per second
+burst: maximum number of tokens/requests in a "token bucket" and is initially full
+each request consumes tokens from the token bucket and if the bucket is empty
+when the request is made, the request is rejected
+*/
+type IPRateLimiter struct {
+	ips   map[string]*ipInfo //contains a map of IP address to information about its request pattern
+	mu    *sync.RWMutex
+	limit rate.Limit
+	burst int
+}
+type ipInfo struct {
+	limiter  *rate.Limiter
+	lastSeen time.Time //this is used to delete entries from the "ips" map
+}
+
+func NewIPRateLimiter(limit rate.Limit, burst int) *IPRateLimiter {
+	i := &IPRateLimiter{
+		ips:   make(map[string]*ipInfo),
+		mu:    &sync.RWMutex{},
+		limit: limit,
+		burst: burst,
+	}
+
+	return i
+}
+
+// AddIP creates a new rate limiter and adds it to the ips map,
+// using the IP address as the key
+func (i *IPRateLimiter) AddIP(ip string) *rate.Limiter {
+	var onlyOnce sync.Once
+
+	var limiter *rate.Limiter
+	onlyOnce.Do(func() {
```

Comment:
@bbengfort - I think this takes care of the double checked locking but let me know if this is not correct.

Reply:
It's possible that I'm misinterpreting this but it seems that the `GetLimiter` function grabs an Rlock, then unlock and calls AddIP; the `AddIP` function creates `onlyOnce` inside of the function; so that mutex is locally scoped. Inside of the function, `onlyOnce` will only be executed once, however multiple concurrent calls to `AddIP` will each have their own `onlyOnce` and will not be synchronized; so it will be overwrritten if the calls happen in quick succession (which may be what is happening in the tests). 

Double checked locking would be:

```
func (i *IPRateLimiter) AddIP(ip string) (*rate.Limiter) {
	i.mu.Lock()
	defer i.mu.Unlock()

	// Here is the second check, e.g. the double check
	if ipInfo, exists := i.ips[ip]; exists {
		return ipInfo.limiter
	}

	// Otherwise the condition from the RLock is still true, so create the limiter 
	limiter := rate.NewLimiter(i.limit, i.burst)
	i.ips[ip] = &ipInfo{limiter, time.Now()}
	return limiter 
}
```

Reply:
understood - so that point behind the double check is to do the check in the `AddIP` method to ensure that ipInfo does not exist before adding the IP address to the map.